el_reward_engine.go

```diff
@@ -0,0 +1,321 @@
+			r.logger.Error().Uint64("slot", slot).Str("relayURL", baseURL).Msg("invalid relay URL, skipping")
+			continue
+		}
+
+		fullURL := fmt.Sprintf("%s%s%d", baseURL, proposerHeaderDeliveredURI, slot)
+		req, err := http.NewRequest(http.MethodGet, fullURL, nil)
+		if err != nil {
+			r.logger.Error().Err(err).Uint64("slot", slot).Str("url", fullURL).Msg("failed to create proposer header delivered request")
+			continue
+		}
+		req.Header.Set("X-API-Key", apiKey)
+		resp, err := r.httpClient.Do(req)
+		if err != nil {
+			r.logger.Error().Err(err).Uint64("slot", slot).Str("url", fullURL).Msgf("http request failed to %s", fullURL)
+			continue
+		}
+		if resp.StatusCode != http.StatusOK {
+			body, _ := io.ReadAll(resp.Body)
+			r.logger.Error().
+				Uint64("slot", slot).
+				Str("url", fullURL).
+				Int("statusCode", resp.StatusCode).
+				Str("body", string(body)).
+				Msg("non-200 response from relay")
+			resp.Body.Close()
+			continue
+		}
+
+		body, err := io.ReadAll(resp.Body)
+		errClose := resp.Body.Close()
+		if err != nil {
+			r.logger.Error().Err(err).Uint64("slot", slot).Str("url", fullURL).Msg("failed to read body from relay")
+			continue
+		}
+		if errClose != nil {
+			r.logger.Warn().Err(errClose).Str("url", fullURL).Msg("error closing response body")
+		}
+
+		var bids []DataHeader
+		if err = json.Unmarshal(body, &bids); err != nil {
+			r.logger.Error().Err(err).Uint64("slot", slot).Str("url", fullURL).Msg("failed to unmarshal data header")
+			continue
+		}
+		if len(bids) == 0 {
+			r.logger.Warn().Uint64("slot", slot).Str("url", fullURL).Msg("no bids received from relay")
+			continue
+		}
+		for i := range bids {
+			bids[i].RelayURL = baseURL
+		}
+
+		relayBids[baseURL] = bids
+	}
+
+	return relayBids
+}
+
+// TODO: Recovery option, move as a standalone script to calculate based on db and endpoints
+func (r *RewardEngine) calculateElRewardInfo(slotStats SlotStatsRecord, groupedBids map[string][]DataHeader) ElRewardInfo {
+	elInfo := ElRewardInfo{
+		ValidatorID:                     slotStats.ValidatorID,
+		AccountID:                       slotStats.AccountID,
+		Slot:                            fmt.Sprintf("%d", slotStats.Slot),
+		SlotUID:                         slotStats.HeaderSlotUID,
+		BlockHash:                       slotStats.PayloadDeliveredBlockHash,
+		SelectedProposerStartTimeUnixMs: slotStats.HeaderStartTimeUnixMs,
+		IsPayloadReceived:               slotStats.PayloadSucceeded,
+		IsProxyWin:                      slotStats.HeaderDeliveredBlockHash == slotStats.PayloadDeliveredBlockHash,
+		Bids:                            make(map[string][]DataHeader),
+		ElRewardIncreaseWei:             big.NewInt(0),
+		OnchainBidDeliveredRelay:        make([]string, 0, 10),
+		SecondHighestBidDeliveredRelay:  make([]string, 0, 10),
+	}
+
+	var matchedBids []DataHeader
+	var errBuf []string
+	for relayURL, bids := range groupedBids {
+		elInfo.Bids[relayURL] = bids
```

Comment:
we shouldn't need allbids only bids that match otherwise this data is going to be very large due to bloxroute not filtering completely